File: backend/rating/src/main/java/com/commonsengine/rating/infrastructure/persistence/RatingEntity.java
package com.commonsengine.rating.infrastructure.persistence;

import com.commonsengine.rating.domain.Model.RatingDirection;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.time.Instant;

/**
 * 评价 JPA 实体。
 *
 * <p>双向评价（消费者↔劳动者），反惩罚性设计——评价不挂钩接单资格。
 * tags 以分号分隔的字符串存储，如 "POLITE;PUNCTUAL"。
 *
 * <p>JPA 要求无参构造器与可变字段，故使用普通 Java 类。
 */
@Entity
@Table(name = "ratings")
public class RatingEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "rating_id", nullable = false, length = 36)
    private String ratingId;

    @Column(name = "transaction_id", nullable = false, length = 36)
    private String transactionId;

    @Column(name = "rater_id", nullable = false, length = 36)
    private String raterId;

    @Column(name = "ratee_id", nullable = false, length = 36)
    private String rateeId;

    @Enumerated(EnumType.STRING)
    @Column(name = "direction", nullable = false, length = 30)
    private RatingDirection direction;

    @Column(name = "score", nullable = false)
    private int score;

    /** 标签以分号分隔存储，如 "POLITE;PUNCTUAL" */
    @Column(name = "tags")
    private String tags;

    @Column(name = "comment", columnDefinition = "TEXT")
    private String comment;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    /** JPA 要求的无参构造器 */
    public RatingEntity() {
    }

    // ── getters / setters ──────────────────────────────

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getRatingId() {
        return ratingId;
    }

    public void setRatingId(String ratingId) {
        this.ratingId = ratingId;
    }

    public String getTransactionId() {
        return transactionId;
    }

    public void setTransactionId(String transactionId) {
        this.transactionId = transactionId;
    }

    public String getRaterId() {
        return raterId;
    }

    public void setRaterId(String raterId) {
        this.raterId = raterId;
    }

    public String getRateeId() {
        return rateeId;
    }

    public void setRateeId(String rateeId) {
        this.rateeId = rateeId;
    }

    public RatingDirection getDirection() {
        return direction;
    }

    public void setDirection(RatingDirection direction) {
        this.direction = direction;
    }

    public int getScore() {
        return score;
    }

    public void setScore(int score) {
        this.score = score;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }
}

File: backend/rating/src/main/kotlin/com/commonsengine/rating/infrastructure/persistence/RatingEntity.kt
package com.commonsengine.rating.infrastructure.persistence

import com.commonsengine.rating.domain.Rating
import com.commonsengine.rating.domain.RatingDirection
import com.commonsengine.rating.domain.RatingId
import com.commonsengine.rating.domain.RatingTag
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.Instant

/**
 * 评价 JPA 实体
 *
 * 双向评价（消费者↔劳动者），反惩罚性设计——评价不挂钩接单资格。
 * tags: Set<RatingTag> 以分号分隔的字符串存储。
 */
@Entity
@Table(name = "ratings")
class RatingEntity(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(name = "rating_id", nullable = false, length = 36)
    val ratingId: String,

    @Column(name = "transaction_id", nullable = false, length = 36)
    val transactionId: String,

    @Column(name = "rater_id", nullable = false, length = 36)
    val raterId: String,

    @Column(name = "ratee_id", nullable = false, length = 36)
    val rateeId: String,

    @Enumerated(EnumType.STRING)
    @Column(name = "direction", nullable = false, length = 30)
    val direction: RatingDirection,

    @Column(name = "score", nullable = false)
    val score: Int,

    /** 标签以分号分隔存储，如 "POLITE;PUNCTUAL" */
    @Column(name = "tags")
    val tags: String? = null,

    @Column(name = "comment", columnDefinition = "TEXT")
    val comment: String? = null,

    @Column(name = "created_at", nullable = false)
    val createdAt: Instant,
)

// ── Entity ↔ Domain 映射 ──────────────────────────────

/** Entity → 领域模型 */
fun RatingEntity.toDomain(): Rating = Rating(
    id = RatingId(ratingId),
    transactionId = transactionId,
    raterId = raterId,
    rateeId = rateeId,
    direction = direction,
    score = score,
    tags = tags?.let { parseTags(it) } ?: emptySet(),
    comment = comment,
    createdAt = createdAt,
)

/** 领域模型 → Entity */
fun Rating.toEntity(): RatingEntity = RatingEntity(
    ratingId = id.value,
    transactionId = transactionId,
    raterId = raterId,
    rateeId = rateeId,
    direction = direction,
    score = score,
    tags = if (tags.isEmpty()) null else tags.joinToString(";") { it.name },
    comment = comment,
    createdAt = createdAt,
)

/** 解析分号分隔的标签字符串 */
private fun parseTags(s: String): Set<RatingTag> =
    s.split(";")
        .filter { it.isNotBlank() }
        .mapNotNull { name -> runCatching { RatingTag.valueOf(name.trim()) }.getOrNull() }
        .toSet()
